src/Common/PODArray.h

```diff
@@ -428,33 +432,54 @@ class PODArray : public PODArrayBase<sizeof(T), initial_bytes, TAllocator, pad_r
     void insertSmallAllowReadWriteOverflow15(It1 from_begin, It2 from_end, TAllocatorParams &&... allocator_params)
     {
         static_assert(pad_right_ >= 15);
+        static_assert(sizeof(T) == sizeof(*from_begin));
         insertPrepare(from_begin, from_end, std::forward<TAllocatorParams>(allocator_params)...);
         size_t bytes_to_copy = this->byte_size(from_end - from_begin);
         memcpySmallAllowReadWriteOverflow15(this->c_end, reinterpret_cast<const void *>(&*from_begin), bytes_to_copy);
         this->c_end += bytes_to_copy;
     }
 
+    /// Do not insert into the array a piece of itself. Because with the resize, the iterators on themselves can be invalidated.
     template <typename It1, typename It2>
     void insert(iterator it, It1 from_begin, It2 from_end)
     {
+        size_t position [[maybe_unused]] = it - begin();
         size_t bytes_to_copy = this->byte_size(from_end - from_begin);
-        size_t bytes_to_move = (end() - it) * sizeof(T);
+        size_t bytes_to_move = this->byte_size(end() - it);
 
         insertPrepare(from_begin, from_end);
 
         if (unlikely(bytes_to_move))
             memcpy(this->c_end + bytes_to_copy - bytes_to_move, this->c_end - bytes_to_move, bytes_to_move);
 
-        memcpy(this->c_end - bytes_to_move, reinterpret_cast<const void *>(&*from_begin), bytes_to_copy);
+        if constexpr (std::is_same_v<T, std::decay_t<decltype(*from_begin)>>)
+        {
+            memcpy(this->c_end - bytes_to_move, reinterpret_cast<const void *>(&*from_begin), bytes_to_copy);
+        }
+        else
+        {
+            it = begin() + position;
+            for (auto from_it = from_begin; from_it != from_end; ++from_it, ++it)
+                new (&*it) T(*from_it);
+        }
+
         this->c_end += bytes_to_copy;
     }
 
     template <typename It1, typename It2>
     void insert_assume_reserved(It1 from_begin, It2 from_end)
     {
-        size_t bytes_to_copy = this->byte_size(from_end - from_begin);
-        memcpy(this->c_end, reinterpret_cast<const void *>(&*from_begin), bytes_to_copy);
-        this->c_end += bytes_to_copy;
+        if constexpr (std::is_same_v<T, std::decay_t<decltype(*from_begin)>>)
+        {
+            size_t bytes_to_copy = this->byte_size(from_end - from_begin);
+            memcpy(this->c_end, reinterpret_cast<const void *>(&*from_begin), bytes_to_copy);
+            this->c_end += bytes_to_copy;
+        }
+        else
```

Comment:
Apparently this is the method that slows down even w/o resize changes.
https://clickhouse-test-reports.s3.yandex.net/13147/52146218302e7ab2ee6bd7c1172fd7e9fa32f5bc/performance_comparison/report.html#fail1

Where in our code are the different types used? If I add static assert, I only find some places with `char`/`char8_t` mismatch, and no types of different size.

Reply:
@alexey-milovidov

Reply:
Our code does not use different types, but it was found in unrelated unit test by @Enmk.

Reply:
char, signed char, unsigned char, uint8_t should continue to use memcpy.

Reply:
Maybe
is_integral_v and same sizeof is also Ok for memcpy.

I will add this condition to the original PR.

Reply:
If it's test-only, maybe we should change the test to use a loop with push_back to avoid test-only code? And add a static assert instead.

Reply:
It's natural to write vec{1, 2, 3} - initializer list of int.
But without this change it will be very error prone. Even if we check sizeof in static_assert, it will fail for float.

Reply:
https://github.com/ClickHouse/ClickHouse/pull/12278/commits/6f8a274ba04fdec65f168ebe63f1acd715bdbd72

Reply:
Very strange, my fix did not help.